Survey/Survey/Sources/Data/NetworkAPI/Core/RequestConfiguration.swift

```diff
@@ -0,0 +1,57 @@
+//
+//  RequestConfiguration.swift
+//  Survey
+//
+//  Created by Taher on 4/11/22.
+//
+
+import Foundation
+
+protocol RequestConfiguration {
+
+    var baseURL: String { get }
+    var endpoint: String { get }
+    var url: URL? { get }
+    var method: URLRequest.HttpMethod { get }
+    var urlRequest: URLRequest? { get }
+    var parameters: Parameters? { get }
+    var retryCount: Int { get }
+    var timeOutValue: Int { get }
+    var contentType: String { get }
+    var contentTypeValue: String { get }
+}
+
+extension RequestConfiguration {
+
+    var url: URL? {
+        URL(string: baseURL)?.appendingPathComponent(endpoint)
+    }
+
+    var contentType: String {
+        "Content-Type"
+    }
+
+    var contentTypeValue: String {
+        "application/json"
+    }
+
+    var urlRequest: URLRequest? {
+            guard let requestUrl = url else {
+                NSLog("Invalid URL given!")
+                return nil
+            }
+            var request = URLRequest(url: requestUrl, timeoutInterval: TimeInterval(timeOutValue))
+            request.httpMethod = method.current
+            request.addValue(contentTypeValue, forHTTPHeaderField: contentType)
+            request.httpBody = nil
+
+            if let body = parameters {
+                do {
+                    request.httpBody = try JSONSerialization.data(withJSONObject: body, options: .fragmentsAllowed)
+                } catch {
+                    NSLog(error.localizedDescription)
+                }
+            }
+            return request
+    }
+}
```

Comment:
- If our Network layer always expect JSON response, and we don't have any other header value, IMO it's fine not to define `contentType` and `contentTypeValue` in the `RequestConfiguration`, just set it directly to every request:
  ```
  request.addValue("application/json", forHTTPHeaderField: "Content-Type")
  ```
  If you want to make it more general or later need to add more header fields, consider put all those in one place, can be named `headers` instead of making every header field to a pair of variables (`headerA` + `headerAValue`, `headerB` + `headerBValue`, ...):
  ```
  typealias HTTPHeaders = [String: String] <- not strictly necessary, just more readable

  protocol RequestConfiguration {
        // ...
        var headers: HTTPHeaders { get }
  }

  extension RequestConfiguration {

      var headers: [String: String] { ["Content-Type": "application/json"] }
      
      var urlRequest: URLRequest? {
          // ...
          headers.forEach { request.addValue($0.key, forHTTPHeaderField: $0.value) }
      }
  }
  ```
- `URLRequest`'s `httpBody` is already `nil` by default, so we don't have to set it `nil` first, also we can make use of `try?` instead of `do catch`:
  ```
  if let parameters = parameters,
    let data = try? JSONSerialization.data(withJSONObject: parameters, options: .fragmentsAllowed) }
      request.httpBody = data
  }
  ```

Reply:
nice suggestion .. updated in this [commit](https://github.com/Shayokh144/ic-survey-swiftui-combine/pull/43/commits/08e91693e844c6ce389503dced572cfcc5f17cea)